File: cpp/static_candidates/RawRating.cpp
#include "RawRating.h"
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

RawRating::~RawRating() {
	if ( talley ) {
		delete [] talley;
	}
}

inline int ratingCleanse( int r ) {
	return (( r == NO_VOTE ) ? 0 : r );
}

int RawRating::voteRating( int* rating ) {
	int i;
	for ( i = 0; i < numc; i++ ) {
		talley[i] += ratingCleanse( rating[i] );
	}
	return 0;
}

int RawRating::getWinners( int* choiceIndecies ) {
	long max = LONG_MIN;
	int i;
	int numWinners;
	for ( i = 0; i < numc; i++ ) {
		if ( talley[i] > max ) {
			choiceIndecies[0] = i;
			numWinners = 1;
			max = talley[i];
		} else if ( talley[i] == max ) {
			choiceIndecies[numWinners] = i;
			numWinners++;
		}
	}
	return numWinners;
}

int RawRating::write( int fd ) {
	return 0;
}

int RawRating::read( int fd ) {
	return 0;
}

void RawRating::htmlSummary( FILE* cur, char** names ) {
	long max;
	bool* in = new bool[numc];
	int maxi,i;
	for ( i = 0; i < numc; i++ ) {
		in[i] = true;
	}
	if ( names ) {
		fprintf( cur, "<table border=\"1\"><tr><th></th><th>Rating Summation</th></tr>\n" );
	} else {
		fprintf( cur, "<table border=\"1\"><tr><th>Choice Index</th><th>Rating Summation</th></tr>\n" );
	}
	while ( true ) {
		bool done;
		max = LONG_MIN;
		done = true;
		for ( i = 0; i < numc; i++ ) {
			if ( in[i] && talley[i] > max ) {
				done = false;
				maxi = i;
				max = talley[i];
			}
		}
		if ( done ) break;
		i = maxi;
		in[i] = false;
		if ( names ) {
			fprintf( cur, "<tr><td>%s</td><td>%ld</td></tr>\n", names[i], talley[i] );
		} else {
			fprintf( cur, "<tr><td>%d</td><td>%ld</td></tr>\n", i+1, talley[i] );
		}
	}
	fprintf( cur, "</table>\n" );
	delete [] in;
}

void RawRating::print( FILE* fout, char** names ) {
	for ( int i = 0; i < numc; i++ ) {
		if ( names ) {
			fprintf(fout,"%s\t%ld\n", names[i], talley[i] );
		} else {
			fprintf(fout,"%ld\n", talley[i] );
		}
	}
	fflush(fout);
}

void RawRating::squashToAcceptance( int* rating ) {
	for ( int i = 0; i < numc; i++ ) {
		if ( rating[i] >= 0 ) {
			rating[i] = 1;
		} else {
			rating[i] = 0;
		}
	}
}

VotingSystem* RawRating::newRawRating( int numc ) {
	return new RawRating( numc );
}

File: java/org/bolson/vote/RawRating.java
package org.bolson.vote;

/**
 * Raw summation of voter's ratings.
 * External limits must be established.
 @author Brian Olson
 */
public class RawRating extends RatedVotingSystem {
	public RawRating( int numCandidates ) {
		super( numCandidates );
		talley = new double[numc];
		for ( int i = 0; i < numc; i++ ) {
			talley[i] = 0;
		}
	}
	boolean normalize = false;
	boolean normalizeL2 = false;
	boolean stretch = false;
	
	public String name() {
		return "Raw Rating Summation";
	}
	
	public int voteRating( int rating[] )  {
		int i;
		for ( i = 0; i < numc; i++ ) if ( rating[i] != NO_VOTE ) {
			talley[i] += rating[i];
		}
		return 0;
	}
	public int voteRating( float rating[] )  {
		int i;
		for ( i = 0; i < numc; i++ ) if ( !Float.isNaN(rating[i]) ) {
			talley[i] += rating[i];
		}
		return 0;
	}
	public int voteRating( double rating[] )  {
		int i;
		for ( i = 0; i < numc; i++ ) if ( !Double.isNaN(rating[i]) ) {
			talley[i] += rating[i];
		}
		return 0;
	}
	
	public int[] getWinners() {
		double max = Double.NEGATIVE_INFINITY;
		int i;
		int numWinners = 0;
		int choiceIndecies[] = new int[numc];
		for ( i = 0; i < numc; i++ ) {
			if ( talley[i] > max ) {
				choiceIndecies[0] = i;
				numWinners = 1;
				max = talley[i];
			} else if ( talley[i] == max ) {
				choiceIndecies[numWinners] = i;
				numWinners++;
			}
		}
		int winners[] = new int[numWinners];
		for ( i = 0; i < numWinners; i++ ) {
			winners[i] = choiceIndecies[i];
		}
		return winners;
	}
	
	
	void squashToAcceptance( int rating[] ) {
		for ( int i = 0; i < numc; i++ ) {
			if ( rating[i] >= 0 ) {
				rating[i] = 1;
			} else {
				rating[i] = 0;
			}
		}
	}
	
	protected double talley[];
	
	public String toString() {
		StringBuffer sb = new StringBuffer();
		for ( int i = 0; i < talley.length; i++ ) {
			sb.append( talley[i] ).append('\n');
		}
		return sb.toString();
	}
	
	public String toString( String names[] ) {
		StringBuffer sb = new StringBuffer();
		for ( int i = 0; i < talley.length; i++ ) {
			sb.append( names[i] ).append('\t');
			sb.append( talley[i] ).append('\n');
		}
		return sb.toString();
	}
	
	public String htmlSummary( String names[] ) {
		StringBuffer sb;
		double max;
		boolean in[] = new boolean[numc];
		int maxi = 0,i;
		for ( i = 0; i < numc; i++ ) {
			in[i] = true;
		}
		if ( names != null ) {
			sb = new StringBuffer( "<table border=\"1\"><tr><th></th><th>Raw Rating Summation</th></tr>\n" );
		} else {
			sb = new StringBuffer( "<table border=\"1\"><tr><th>Choice Index</th><th>Raw Rating Summation</th></tr>\n" );
		}
		while ( true ) {
			boolean done;
			max = Double.NEGATIVE_INFINITY;
			done = true;
			for ( i = 0; i < numc; i++ ) {
				if ( in[i] && (talley[i] > max) ) {
					done = false;
					maxi = i;
					max = talley[i];
				}
			}
			if ( done ) {
				//sb.append(max);
				break;
			}
			i = maxi;
			in[i] = false;
			sb.append( "<tr><td>" );
			if ( names != null ) {
				sb.append( names[i] );
			} else {
				sb.append( i+1 );
			}
			sb.append( "</td><td>" );
			sb.append( talley[i] );
			sb.append( "</td></tr>\n" );
		}
		sb.append( "</table>\n" );
		/*for ( i = 0; i < numc; i++ ) {
		 sb.append( in[i] ).append(' ').append( talley[i] ).append(", ");
		 }*/
		return sb.toString();
	}
}
